broker/src/main/java/io/zeebe/broker/system/partitions/ZeebePartition.java

```diff
@@ -710,6 +663,57 @@ protected void handleFailure(final Exception failure) {
     onInstallFailure();
   }
 
+  private void addClosingStep(final String name, final AsyncClosable closable) {
+    closingSteps.add(new ClosingStep(name, closable));
+  }
+
+  private CompletableActorFuture<Void> closePartition() {
+    // this method may be called concurrently when the actor is closed
+    final var closingStepsInReverseOrder = new ArrayList<>(closingSteps);
+    Collections.reverse(closingStepsInReverseOrder);
+
+    final var closingPartitionFuture = new CompletableActorFuture<Void>();
+    stepByStepClosing(closingPartitionFuture, closingStepsInReverseOrder);
+
+    return closingPartitionFuture;
+  }
+
+  private void stepByStepClosing(
+      final CompletableActorFuture<Void> closingFuture, final List<ClosingStep> actorsToClose) {
+    if (actorsToClose.isEmpty()) {
+      closingFuture.complete(null);
+      return;
+    }
+
+    final ClosingStep closingStep = actorsToClose.remove(0);
```

Comment:
We remove now twice from two different lists ? :sweat_smile:

Reply:
Yes. We remove the closed item from both lists.